python/python/lance/util.py

```diff
@@ -40,3 +45,71 @@ def sanitize_ts(ts: ts_types) -> datetime:
     elif not isinstance(ts, datetime):
         raise TypeError(f"Unrecognized version timestamp {ts} of type {type(ts)}")
     return ts
+
+
+class KMeans:
+    def __init__(self, k: int, metric_type: str = "l2", max_iters: int = 50):
+        """Create a KMeans model.
+
+        Parameters
+        ----------
+        k: int
+            The number of clusters to create.
+        metric_type: str, default="l2"
+            The metric to use for calculating distances between vectors.
+            Supported distance metrics: "l2", "cosine", "dot"
+        max_iters: int
+            The maximum number of iterations to run the KMeans algorithm. Default: 50.
+        """
+        self.k = k
+        self._metric_type = metric_type
+        self._kmeans = _KMeans(k, metric_type, max_iters=max_iters)
+
+    def __repr__(self) -> str:
+        return f"lance.KMeans(k={self.k}, metric_type={self._metric_type})"
+
+    @property
+    def centroids(self) -> Optional[pa.FixedSizeListArray]:
+        """Returns the centroids of the model."""
+        ret = self._kmeans.centroids()
+        return ret
+
+    def _to_fixed_size_list(self, data: pa.Array) -> pa.FixedSizeListArray:
+        if isinstance(data, pa.FixedSizeListArray):
+            if data.value_type != pa.float32():
+                raise ValueError(f"Array must be float32 type, got: {data.value_type}")
+            return data
+        elif isinstance(data, pa.FixedShapeTensorArray):
+            return pa.FixedSizeListArray.from_arrays(data.storage(), data.shape[1])
+        elif isinstance(data, np.ndarray):
+            if len(data.shape) != 2:
+                raise ValueError(
+                    f"Numpy array must be a 2-D array, got {len(data.shape)}-D"
+                )
+            elif data.dtype != np.float32:
+                raise ValueError(f"Numpy array must be float32 type, got: {data.dtype}")
+
+            inner_arr = pa.array(data.reshape((-1,)), type=pa.float32())
+            return pa.FixedSizeListArray.from_arrays(inner_arr, data.shape[1])
+        else:
+            raise ValueError("Data must be a FixedSizeListArray, Tensor or numpy array")
```

Comment:
This sort of normalization / cleaning routine probably doesn't belong on this class in the long run but maybe ok for now until fixed size arrays are needed elsewhere.